### PR TITLE
Add flask tests

### PR DESCRIPTION
**Pull request by asvrada**

### CODE REVIEW

**Comment by asvrada** [comment]
When running the Docker image by using `docker build . -t flask && docker run -it flask` then run `pytest` in docker container, all tests would pass.

When running it as Mystikos app by using `make tests`, it would fail with the error message:
"""
ImportError while loading conftest '/app/flask/tests/conftest.py'.
tests/conftest.py:9: in <module>
    import flask
src/flask/__init__.py:7: in <module>
    from .app import Flask as Flask
src/flask/app.py:28: in <module>
    from . import cli
src/flask/cli.py:18: in <module>
    from .helpers import get_debug_flag
src/flask/helpers.py:23: in <module>
    from .signals import message_flashed
src/flask/signals.py:4: in <module>
    from blinker import Namespace
/usr/local/lib/python3.9/site-packages/blinker/__init__.py:1: in <module>
    from blinker.base import (
E     File "/usr/local/lib/python3.9/site-packages/blinker/base.py", line 145
E       """Connect the decorated function as a receiver for *sender*.
E       ^
E   SyntaxError: invalid escape sequence \*
"""

This is caused by the usage of `\*` Python library Blinker

> Will investigate why 1) Running in docker doesn't encounter this problem, 2) This is already fixed in the latest Blinker, why we are using an older version here

After manually removing all such occurrence in appdir, `make tests` would actually run the flask tests, and only 1 test related to subprocess would fail
```
              self.pid = _posixsubprocess.fork_exec(
                        args, executable_list,
                        close_fds, tuple(sorted(map(int, fds_to_keep))),
                        cwd, env_list,
                        p2cread, p2cwrite, c2pread, c2pwrite,
                        errread, errwrite,
                        errpipe_read, errpipe_write,
                        restore_signals, start_new_session,
                        gid, gids, uid, umask,
                        preexec_fn)
E                       OSError: [Errno 95] Not supported

/usr/local/lib/python3.9/subprocess.py:1754: OSError
=========================== short test summary info ============================
FAILED tests/test_instance_config.py::test_egg_installed_paths - OSError: [Er...
============= 1 failed, 473 passed, 1 skipped, 3 xpassed in 8.23s ==============
```

**Comment by vtikoo** [comment]
@asvrada If flask is using subprocess/fork, you need to opt into the pseudo fork support in Mystikos. Its disabled by default. See - https://github.com/deislabs/mystikos/blob/main/doc/design/fork.md#design.

Recommend using the `pseudo_wait_for_exit_exec` fork mode.

**Comment by asvrada** [comment]
> @asvrada If flask is using subprocess/fork, you need to opt into the pseudo fork support in Mystikos. Its disabled by default. See - https://github.com/deislabs/mystikos/blob/main/doc/design/fork.md#design.
> 
> Recommend using the `pseudo_wait_for_exit_exec` fork mode.

Is this by design? i.e. code that uses fork would fail as the current state of Mystikos, but expected to work with  `pseudo_wait_for_exit_exec` fork mode enabled?

**Comment by vtikoo** [comment]
> 
> Is this by design? i.e. code that uses fork would fail as the current state of Mystikos, but expected to work with `pseudo_wait_for_exit_exec` fork mode enabled?

Yes, as there are limitations(the design doc goes into more detail) to supporting fork in a single enclave, it requires explicit opt in via command line or config.json. The default behavior is to return a not supported error(ENOTSUP).

**Comment by asvrada** [comment]
The test still fails with either fork mode

```
pseudo_wait_for_exit_exec or pseudo

___________________________ test_egg_installed_paths ___________________________

install_egg = <function install_egg.<locals>.inner at 0x2965baf0>
modules_tmpdir = local('/tmp/pytest-of-root/pytest-0/test_egg_installed_paths0/modules_tmpdir')
modules_tmpdir_prefix = local('/tmp/pytest-of-root/pytest-0/test_egg_installed_paths0/modules_tmpdir')

    def test_egg_installed_paths(install_egg, modules_tmpdir, modules_tmpdir_prefix):
        modules_tmpdir.mkdir("site_egg").join("__init__.py").write(
            "import flask\n\napp = flask.Flask(__name__)"
        )
>       install_egg("site_egg")

tests/test_instance_config.py:112: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/conftest.py:182: in inner
    subprocess.check_call(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

popenargs = (['/usr/local/bin/python', 'setup.py', 'bdist_egg'],)
kwargs = {'cwd': '/tmp/pytest-of-root/pytest-0/test_egg_installed_paths0/modules_tmpdir'}
retcode = 1, cmd = ['/usr/local/bin/python', 'setup.py', 'bdist_egg']

    def check_call(*popenargs, **kwargs):
        """Run command with arguments.  Wait for command to complete.  If
        the exit code was zero then return, otherwise raise
        CalledProcessError.  The CalledProcessError object will have the
        return code in the returncode attribute.
    
        The arguments are the same as for the call function.  Example:
    
        check_call(["ls", "-l"])
        """
        retcode = call(*popenargs, **kwargs)
        if retcode:
            cmd = kwargs.get("args")
            if cmd is None:
                cmd = popenargs[0]
>           raise CalledProcessError(retcode, cmd)
E           subprocess.CalledProcessError: Command '['/usr/local/bin/python', 'setup.py', 'bdist_egg']' returned non-zero exit status 1.

/usr/local/lib/python3.9/subprocess.py:373: CalledProcessError
=========================== short test summary info ============================
FAILED tests/test_instance_config.py::test_egg_installed_paths - subprocess.C...
============= 1 failed, 473 passed, 1 skipped, 3 xpassed in 10.34s =============

```

File: tests/python_flask_tests/config.json [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+    "version": "0.1",
+    "Debug": 1,
+    "ProductID": 1,
+    "SecurityVersion": 1,
+    "MemorySize": "256m",
+    "CurrentWorkingDirectory": "/app/flask",
+    "ApplicationPath": "/usr/local/bin/python",
+    "ApplicationParameters": ["/usr/local/bin/pytest"],
+    "ForkMode": "pseudo_wait_for_exit_exec"
+}
```
Comment (bodzhang) [inline tests/python_flask_tests/config.json:11]: Pytest use multi-process. For Mystikos Multi-process support, "NoBrk" should be set (to true)

**Comment by bodzhang** [comment]
> When running the Docker image by using `docker build . -t flask && docker run -it flask` then run `pytest` in docker container, all tests would pass.
> 
> When running it as Mystikos app by using `make tests`, it would fail with the error message:
> 
> ```
> ImportError while loading conftest '/app/flask/tests/conftest.py'.
> tests/conftest.py:9: in <module>
>     import flask
> src/flask/__init__.py:7: in <module>
>     from .app import Flask as Flask
> src/flask/app.py:28: in <module>
>     from . import cli
> src/flask/cli.py:18: in <module>
>     from .helpers import get_debug_flag
> src/flask/helpers.py:23: in <module>
>     from .signals import message_flashed
> src/flask/signals.py:4: in <module>
>     from blinker import Namespace
> /usr/local/lib/python3.9/site-packages/blinker/__init__.py:1: in <module>
>     from blinker.base import (
> E     File "/usr/local/lib/python3.9/site-packages/blinker/base.py", line 145
> E       """Connect the decorated function as a receiver for *sender*.
> E       ^
> E   SyntaxError: invalid escape sequence \*
> ```
> 
> This is caused by the usage of `\*` in a Python library Blinker (https://github.com/jek/blinker/blob/master/blinker/base.py#L96)
> 
> > Will investigate why 1) Running in docker doesn't encounter this problem, 2) Blinker lib already [fixed this issue](https://github.com/jek/blinker/commit/194e17a20008bd778714182ac191ad654074f81c), why we are using an older version here
> 
> After manually removing all such occurrences in `appdir/usr/local/lib/python3.9/site-packages/blinker/base.py`, `make tests` would run the flask tests, and only 1 test related to subprocess/fork would fail
> 
> ```
>               self.pid = _posixsubprocess.fork_exec(
>                         args, executable_list,
>                         close_fds, tuple(sorted(map(int, fds_to_keep))),
>                         cwd, env_list,
>                         p2cread, p2cwrite, c2pread, c2pwrite,
>                         errread, errwrite,
>                         errpipe_read, errpipe_write,
>                         restore_signals, start_new_session,
>                         gid, gids, uid, umask,
>                         preexec_fn)
> E                       OSError: [Errno 95] Not supported
> 
> /usr/local/lib/python3.9/subprocess.py:1754: OSError
> =========================== short test summary info ============================
> FAILED tests/test_instance_config.py::test_egg_installed_paths - OSError: [Er...
> ============= 1 failed, 473 passed, 1 skipped, 3 xpassed in 8.23s ==============
> ```

Does the native container have some environment variables the Mystikos config.json file didn't capture?

File: tests/Makefile [hunk 2]
```diff
@@ -114,6 +114,7 @@ DIRS += dotnet-sos
 DIRS += tkillself
 DIRS += thread_abort
 DIRS += synccall
+DIRS += python_flask_tests
```
Comment (bodzhang) [inline tests/Makefile:117]: Running an existing test suite of a python module qualifies as a "solution". Suggest to move to solutions directory.

**Comment by bodzhang** [comment]
> The test still fails with either fork mode
> 
> ```
> pseudo_wait_for_exit_exec or pseudo
> 
> ___________________________ test_egg_installed_paths ___________________________
> 
> install_egg = <function install_egg.<locals>.inner at 0x2965baf0>
> modules_tmpdir = local('/tmp/pytest-of-root/pytest-0/test_egg_installed_paths0/modules_tmpdir')
> modules_tmpdir_prefix = local('/tmp/pytest-of-root/pytest-0/test_egg_installed_paths0/modules_tmpdir')
> 
>     def test_egg_installed_paths(install_egg, modules_tmpdir, modules_tmpdir_prefix):
>         modules_tmpdir.mkdir("site_egg").join("__init__.py").write(
>             "import flask\n\napp = flask.Flask(__name__)"
>         )
> >       install_egg("site_egg")
> 
> tests/test_instance_config.py:112: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> tests/conftest.py:182: in inner
>     subprocess.check_call(
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> 
> popenargs = (['/usr/local/bin/python', 'setup.py', 'bdist_egg'],)
> kwargs = {'cwd': '/tmp/pytest-of-root/pytest-0/test_egg_installed_paths0/modules_tmpdir'}
> retcode = 1, cmd = ['/usr/local/bin/python', 'setup.py', 'bdist_egg']
> 
>     def check_call(*popenargs, **kwargs):
>         """Run command with arguments.  Wait for command to complete.  If
>         the exit code was zero then return, otherwise raise
>         CalledProcessError.  The CalledProcessError object will have the
>         return code in the returncode attribute.
>     
>         The arguments are the same as for the call function.  Example:
>     
>         check_call(["ls", "-l"])
>         """
>         retcode = call(*popenargs, **kwargs)
>         if retcode:
>             cmd = kwargs.get("args")
>             if cmd is None:
>                 cmd = popenargs[0]
> >           raise CalledProcessError(retcode, cmd)
> E           subprocess.CalledProcessError: Command '['/usr/local/bin/python', 'setup.py', 'bdist_egg']' returned non-zero exit status 1.
> 
> /usr/local/lib/python3.9/subprocess.py:373: CalledProcessError
> =========================== short test summary info ============================
> FAILED tests/test_instance_config.py::test_egg_installed_paths - subprocess.C...
> ============= 1 failed, 473 passed, 1 skipped, 3 xpassed in 10.34s =============
> ```

With latest Main branch, STRACE around the time the subprocess shuting down:

...
=== SYS_rename(oldpath="/usr/local/lib/python3.9/site-packages/setuptools/command/__pycache__/bdist_egg.cpython-39.pyc.653094608" newpath="/usr/local/lib/python3.9/site-packages/setuptools/command/__pycache__/bdist_egg.cpython-39.pyc"): pid=102 tid=102
    SYS_rename(): return=0(0): pid=102 tid=102
    SYS_uname(): return=0(0): pid=102 tid=102
=== SYS_write(fd=1 buf=0x2817f230 count=18): pid=102 tid=102
    SYS_write(): return=-EBADF(-9): pid=102 tid=102
=== SYS_write(fd=1 buf=0x2817f230 count=18): pid=102 tid=102
    SYS_write(): return=-EBADF(-9): pid=102 tid=102
=== SYS_write(fd=2 buf=0x2817f430 count=37): pid=102 tid=102
    SYS_write(): return=-EBADF(-9): pid=102 tid=102
=== SYS_writev(fd=2 iov=0x286d6730 iovcnt=2): pid=102 tid=102
    SYS_writev(): return=-EBADF(-9): pid=102 tid=102
=== SYS_write(fd=2 buf=0x2817f430 count=37): pid=102 tid=102
    SYS_write(): return=-EBADF(-9): pid=102 tid=102
=== SYS_write(fd=1 buf=0x2817f230 count=18): pid=102 tid=102
    SYS_write(): return=-EBADF(-9): pid=102 tid=102
=== SYS_write(fd=2 buf=0x2817f430 count=121): pid=102 tid=102
    SYS_write(): return=-EBADF(-9): pid=102 tid=102
=== SYS_write(fd=2 buf=0x2817f430 count=197): pid=102 tid=102
    SYS_write(): return=-EBADF(-9): pid=102 tid=102
=== SYS_write(fd=2 buf=0x2817f430 count=197): pid=102 tid=102
    SYS_write(): return=-EBADF(-9): pid=102 tid=102
=== SYS_write(fd=2 buf=0x2817f430 count=197): pid=102 tid=102
    SYS_write(): return=-EBADF(-9): pid=102 tid=102
=== SYS_write(fd=2 buf=0x2817f430 count=197): pid=102 tid=102
    SYS_write(): return=-EBADF(-9): pid=102 tid=102
=== SYS_write(fd=1 buf=0x2817f230 count=18): pid=102 tid=102
    SYS_write(): return=-EBADF(-9): pid=102 tid=102
=== SYS_write(fd=1 buf=0x2817f230 count=18): pid=102 tid=102
    SYS_write(): return=-EBADF(-9): pid=102 tid=102
=== SYS_munmap(addr=276a2000 length=262144(40000)): pid=102 tid=102
    SYS_munmap(): return=0(0): pid=102 tid=102
=== SYS_rt_sigprocmask(how=0 set=0x287bc370 oldset=0x286d6cd0): pid=102 tid=102
    SYS_rt_sigprocmask(): return=0(0): pid=102 tid=102
=== SYS_rt_sigprocmask(how=2 set=0x286d6cd0 oldset=0): pid=102 tid=102
    SYS_rt_sigprocmask(): return=0(0): pid=102 tid=102
=== SYS_myst_get_fork_info(): pid=102 tid=102
    SYS_myst_get_fork_info(): return=0(0): pid=102 tid=102
=== SYS_exit_group(status=120): pid=102 tid=102
    SYS_exit_group(): return=0(0): pid=102 tid=102
...

Somehow the subprocess that runs {'/usr/local/bin/python', 'setup.py', 'bdist_egg'} failed with error code 120. The only suspious strace is  -EBADF on fd=1/2. Probably need  python debugger to chase down further.

File: tests/Makefile (hunk 2 above)
Comment (asvrada) [reply tests/Makefile:117]: Will move to solution

File: tests/python_flask_tests/config.json (hunk 1 above)
Comment (asvrada) [reply tests/python_flask_tests/config.json:11]: Added "NoBrk": true but the test would still fail

**Comment by bodzhang** [comment]
test_instance_config.py::test_egg_installed_paths failed because of issue #503. The test's main process called dup2 to dup certain files to fd 0/1/2,. dup2() is expected to not copy FD_CLOEXEC flag to the new fds (in this case fd 0/1/2). And in the child process, after execv, fd 0/1/2 are expected to remain open. As pointed out in issue #503, Mystikos current implementation effectively causes FD_CLOEXEC flag to be copied, and the execv flow in the child process closed fd 0/1/2.

**Review by bodzhang — APPROVED** [review]

**Pull request merged**